docs/proposals/node-local-services.md

```diff
@@ -0,0 +1,260 @@
+    - [API changes](#api-changes)
+    - [`kube-proxy` changes](#kube-proxy-changes)
+      - [iptables kernel mode](#iptables-kernel-mode)
+      - [ipvs kernel mode](#ipvs-kernel-mode)
+      - [proxy userland mode](#proxy-userland-mode)
+    - [n-to-1 forwarding](#n-to-1-forwarding)
+  - [Implementation plan](#implementation-plan)
+    - [API](#api)
+    - [kube-proxy](#kube-proxy)
+    - [Documentation](#documentation)
+  - [Future work](#future-work)
+  - [Security considerations](#security-considerations)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+# Node-local services
+
+Author: @therc
+
+Date: July 2016
+
+Status: Design in progress
+
+## Objective
+
+Users should be able to create a service that connects pods to a daemon that is
+running on the same node (and not elsewhere). Possible use cases:
+
+ - monitoring agents and forwarders such as datadog and
+   [tallier](https://github.com/reddit/tallier)
+ - connection poolers and service discovery systems such as
+   [pgbouncer](https://pgbouncer.github.io/) and
+   [synapse](https://github.com/airbnb/synapse)
+ - logging agents such as fluentd
+ - authenticating proxies such as [aws-es-proxy](https://github.com/kopeio/aws-es-proxy),
+   [kube2iam](https://github.com/jtblin/kube2iam) or loasd ([#2209](https://github.com/kubernetes/kubernetes/issues/2209))
+
+At the very least, this new kind of service should support 1:1
+_service_->_daemon_ mappings. For cases like `loasd`, it would be ideal to
+support n:1 mapping: a number of services all getting forwarded to the same
+daemon, with the daemon able to tell which service each connection was meant
+for.
+
+## Background
+
+Thanks to DaemonSets, it is possible to schedule daemon pods on every node (or a
+specific subset if node labels are specified). Local discovery is more
+complicated and currently requires e.g. iptables hacks that bypass Kubernetes
+altogether. This is because existing types of services will pick a random
+service pod when a client establishes a new connection. More often than not, the
+client pod will talk to a service pod running on another machine.
+
+This is not acceptable for some use cases. The daemon might return node-specific
+information downstream or report it upstream to a server. If it were to receive
+traffic from another machine, data returned or forwarded would be inaccurately
+tagged. In other cases, the main concern is reducing cross-node traffic, be it
+for bandwidth or security reasons.
+
+Thus, a new service type is needed that always points to the node-local daemon
+pod.
+
+## Detailed discussion
+
+Node-local services can reuse most of the existing plumbing.
+
+### API changes
+
+To keep changes to a minimum, a new ServiceAffinity, `RequireNodeLocal` is
+added. This strongly implies that node-local services are of `ClusterIP` type
+and point to DaemonSets, both reasonable assumptions to make.
+
+Last, but not least: the current assumption is that service and DaemonSet reside
+in the same namespace.
```

Comment:
The first alternative I can think of would be to have an additional lever that tells the API to create clones of the service in all namespaces. That complicates things quickly, though.

I can see use cases for both global (datadog) and namespace-local (pgbouncer) services. For the most part, this is an issue also for regular services.